### PR TITLE
[INTEL MKL] Add primitive cache for mkl softmax

### PR DESCRIPTION
**Pull request by Solaryee**
Enable primitive reuse in MklSoftmax, please review.

### CODE REVIEW

**Comment by ShengYang1** [comment]
@xiejw  Could you take a look at this PR and give some suggestions? Thank you!

File: tensorflow/core/kernels/mkl_softmax_op.cc [hunk 1]
```diff
@@ -17,21 +17,168 @@ limitations under the License.
...
+
+  struct SoftmaxFwdContext context_;
+  engine cpu_engine_;
+};
+
+template <typename T>
+class MklSoftmaxPrimitiveFactory : public MklPrimitiveFactory<T> {
+ public:
+  static MklSoftmaxPrimitive<T>* Get(const MklSoftmaxParams& fwdParams) {
+    MklSoftmaxPrimitive<T>* softmax_forward = nullptr;
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_softmax_op.cc]: Let's initialize this to the value in line 143 instead of `nullptr`.

File: tensorflow/core/kernels/mkl_softmax_op.cc [hunk 2]
```diff
@@ -107,70 +251,53 @@ class MklSoftmaxOp : public OpKernel {
                          errors::Aborted("Input dims must be <= 5 and >=1"));
           return;
       }
-      // Create softmax memory for src, dst: both are defined in mkl_util.h,
-      // they are wrapper
-      MklDnnData<T> src(&cpu_engine);
-      MklDnnData<T> dst(&cpu_engine);
 
       // If input is in MKL layout, then simply grab input layout; otherwise,
       // construct input Tf layout. For TF layout, although input shape
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_softmax_op.cc]: Please update the comment.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR and sorry for our delay!

File: tensorflow/core/kernels/mkl_softmax_op.cc [hunk 3]
```diff
@@ -17,21 +17,168 @@ limitations under the License.
...
+  struct SoftmaxFwdContext {
+    // MKLDNN memory
+    std::shared_ptr<memory> src_mem;
+    std::shared_ptr<memory> dst_mem;
+
+    // desc & prmitive desc
+    std::shared_ptr<mkldnn::softmax_forward::desc> fwd_desc;
+
+    // memory desc
+    std::shared_ptr<memory::desc> src_md;
```
[the remark this answers is not in the document]
Comment (ShengYang1) [reply tensorflow/core/kernels/mkl_softmax_op.cc]: We use shared_ptr just for easily resource management.

File: tensorflow/core/kernels/mkl_softmax_op.cc [hunk 4]
```diff
@@ -17,21 +17,168 @@ limitations under the License.
...
+};
+
+template <typename T>
+class MklSoftmaxPrimitive : public MklPrimitive {
+ public:
+  explicit MklSoftmaxPrimitive(const MklSoftmaxParams& fwdParams)
+      : cpu_engine_(engine::cpu, 0) {
+    context_.fwd_stream.reset(new stream(stream::kind::eager));
+
+    if (context_.softmax_fwd == nullptr) {
```
[the remark this answers is not in the document]
Comment (ShengYang1) [reply tensorflow/core/kernels/mkl_softmax_op.cc]: MklPrimitive was implemented in another way and this if expression was necessary before. Now it can be safely deleted. Thanks for your correction.

**Comment by ShengYang1** [comment]
@penpornk Changes applied.

**Review by penpornk — APPROVED** [review]
Thank you for the fixes!

File: tensorflow/core/kernels/mkl_softmax_op.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_softmax_op.cc]: If that's the case, would `unique_ptr` also work for you? It should be more efficient than `shared_ptr`.

**Review by penpornk — CHANGES_REQUESTED** [review]
I'm temporarily taking back my approval. I feel a bit weird about having unnecessary `shared_ptr`.

File: tensorflow/core/kernels/mkl_softmax_op.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_softmax_op.cc]: Never mind. I see this in other PRs as well. I think this is more of a design question. I'll bring this up in a sync meeting later.

**Review by penpornk — APPROVED** [review]
